File: workspace/pi_mpi/pi2.f90
	program testpi2
	include 'mpif.h'
	REAL SUM , API
	INTEGER I , NN , NTOT ,IERR, NSIZE , IRANK ,I1,I2
	NN = 1000

	call MPI_INIT(ierr)

	call MPI_COMM_SIZE(MPI_COMM_WORLD,nsize , ierr)
	NTOT = NN*nsize
	call MPI_COMM_RANK(MPI_COMM_WORLD, irank , ierr) 
!	WRITE(*,*) ISIZE,IRANK
	I1 = 1+irank*NN
	I2 = (IRANK+1)*NN
	sum = 0.0 
	DO I = I1 , I2
	  SUM = SUM + 1.0/( 1.0 + ( (REAL(I)-0.5)/REAL(NTOT) ) **2 )
	ENDDO
	API = 4.0*SUM/NTOT
	WRITE(*,*) 'PARTIAL PI =' , API
	CALL MPI_FINALIZE(ierr) 
	END

File: workspace/pi_mpi/pi2.c
#include <stdio.h>
#include <mpi.h>
#define SQR(x) ( (x) * (x) )
 int main(int argc, char *argv[])
/*
   A parallel code calculating segments of the series expansion of PI. */
 {   
        int nsize , irank  , ntot , i1 ,i2 , i , nn ;                                                            
	float sum , api ;
        MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD , &nsize  ) ;
	MPI_Comm_rank(MPI_COMM_WORLD,  &irank ) ;
	nn = 100;
	ntot = nn*nsize;
	sum = 0.0 ;
	i1 = 1 + irank*nn;
	i2 = (irank+1)*nn;
	for (i=i1;i<=i2;i++)
	{ 
	  sum = sum + 1.0/(   1.0 +  SQR( ((float)i-0.5)/ (float)ntot   )  );
	}
	api = 4.0*sum/( (float) ntot );
	printf ("partial pi =%f\n" , api );
	 
	MPI_Finalize();
 }
